File: android/src/main/java/io/linkrunner/capacitor/LinkrunnerPlugin.kt
package io.linkrunner.capacitor

import android.util.Log
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import io.linkrunner.sdk.LinkRunner
import io.linkrunner.sdk.models.request.UserDataRequest
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

/**
 * Capacitor plugin for Linkrunner SDK
 * Provides a bridge between JavaScript/TypeScript and the native Android Linkrunner SDK
 */
@CapacitorPlugin(name = "Linkrunner")
class LinkrunnerPlugin : Plugin() {

    companion object {
        private const val TAG = "LinkrunnerPlugin"
    }

    private lateinit var linkrunner: LinkRunner
    private val coroutineScope = CoroutineScope(Dispatchers.IO)

    override fun load() {
        super.load()
        // Initialize the Linkrunner SDK instance
        linkrunner = LinkRunner.getInstance()
    }

    /**
     * Initialize the Linkrunner SDK
     */
    @PluginMethod
    fun init(call: PluginCall) {
        val token = call.getString("token")
        val secretKey = call.getString("secretKey")
        val keyId = call.getString("keyId")
        val debug = call.getBoolean("debug", false) ?: false
        val packageVersion = call.getString("packageVersion")

        if (token.isNullOrBlank()) {
            call.reject("INVALID_PARAMETER", "Token is required")
            return
        }

        val context = context
        if (context == null) {
            call.reject("CONTEXT_UNAVAILABLE", "Context is not available")
            return
        }

        // Configure SDK for Capacitor platform
        if (!packageVersion.isNullOrEmpty()) {
            LinkRunner.configureSDK("CAPACITOR", packageVersion)
        }

        // Launch coroutine to call suspend function
        coroutineScope.launch {
            try {
                val result = linkrunner.init(
                    context = context,
                    token = token,
                    link = null,
                    source = null,
                    secretKey = secretKey,
                    keyId = keyId,
                    debug = debug
                )

                if (result.isSuccess) {
                    call.resolve()
                } else {
                    val exception = result.exceptionOrNull()
                    Log.e(TAG, "Init failed", exception)
                    call.reject("INIT_FAILED", exception?.message ?: "Initialization failed")
                }
            } catch (e: Exception) {
                Log.e(TAG, "Init failed", e)
                call.reject("INIT_FAILED", e.message ?: "Initialization failed")
            }
        }
    }

    /**
     * Register a new user signup
     */
    @PluginMethod
    fun signup(call: PluginCall) {
        val userDataObj = call.getObject("userData")
        val additionalData = call.getObject("data")

        if (userDataObj == null) {
            call.reject("INVALID_PARAMETER", "userData is required")
            return
        }

        try {
            val userData = convertToUserDataRequest(userDataObj)
            val dataMap = additionalData?.let { convertJSObjectToMap(it) }

            coroutineScope.launch {
                try {
                    val result = linkrunner.signup(userData, dataMap)

                    if (result.isSuccess) {
                        call.resolve()
                    } else {
                        val exception = result.exceptionOrNull()
                        Log.e(TAG, "Signup failed", exception)
                        call.reject("SIGNUP_FAILED", exception?.message ?: "Signup failed")
                    }
                } catch (e: Exception) {
                    Log.e(TAG, "Signup failed", e)
                    call.reject("SIGNUP_FAILED", e.message ?: "Signup failed")
                }
            }
        } catch (e: Exception) {
            Log.e(TAG, "Failed to convert user data", e)
            call.reject("DATA_CONVERSION_FAILED", e.message ?: "Failed to convert user data")
        }
    }

    /**
     * Update user data
     */
    @PluginMethod
    fun setUserData(call: PluginCall) {
        val userDataObj = call.getObject("userData")

        if (userDataObj == null) {
            call.reject("INVALID_PARAMETER", "userData is required")
            return
        }

        try {
            val userData = convertToUserDataRequest(userDataObj)

            coroutineScope.launch {
                try {
                    val result = linkrunner.setUserData(userData)

                    if (result.isSuccess) {
                        call.resolve()
                    } else {
                        val exception = result.exceptionOrNull()
                        Log.e(TAG, "SetUserData failed", exception)
                        call.reject("SET_USER_DATA_FAILED", exception?.message ?: "SetUserData failed")
                    }
                } catch (e: Exception) {
                    Log.e(TAG, "SetUserData failed", e)
                    call.reject("SET_USER_DATA_FAILED", e.message ?: "SetUserData failed")
                }
            }
        } catch (e: Exception) {
            Log.e(TAG, "Failed to convert user data", e)
            call.reject("DATA_CONVERSION_FAILED", e.message ?: "Failed to convert user data")
        }
    }

    /**
     * Convert JSObject to UserDataRequest
     */
    private fun convertToUserDataRequest(jsObject: JSObject): UserDataRequest {
        val id = jsObject.getString("id")
            ?: throw IllegalArgumentException("User id is required")

        return UserDataRequest(
            id = id,
            name = jsObject.getString("name"),
            phone = jsObject.getString("phone"),
            email = jsObject.getString("email"),
            mixpanelDistinctId = jsObject.getString("mixpanel_distinct_id"),
            amplitudeDeviceId = jsObject.getString("amplitude_device_id"),
            posthogDistinctId = jsObject.getString("posthog_distinct_id"),
            brazeDeviceId = jsObject.getString("braze_device_id"),
            gaAppInstanceId = jsObject.getString("ga_app_instance_id"),
            gaSessionId = jsObject.getString("ga_session_id"),
            userCreatedAt = jsObject.getString("user_created_at"),
            isFirstTimeUser = jsObject.getBool("is_first_time_user")
        )
    }

    /**
     * Convert JSObject to Map
     */
    private fun convertJSObjectToMap(jsObject: JSObject): Map<String, Any> {
        val map = mutableMapOf<String, Any>()
        val iterator = jsObject.keys()
        while (iterator.hasNext()) {
            val key = iterator.next()
            val value = jsObject.get(key)
            if (value != null) {
                map[key] = value
            }
        }
        return map
    }

    /**
     * Capture a payment event
     */
    @PluginMethod
    fun capturePayment(call: PluginCall) {
        val userId = call.getString("userId")
        val amount = call.getDouble("amount")
        val paymentId = call.getString("paymentId")
        val typeStr = call.getString("type", "DEFAULT")
        val statusStr = call.getString("status", "PAYMENT_COMPLETED")

        if (userId.isNullOrBlank()) {
            call.reject("INVALID_PARAMETER", "userId is required")
            return
        }

        if (amount == null) {
            call.reject("INVALID_PARAMETER", "amount is required")
            return
        }

        try {
            val type = convertToPaymentType(typeStr)
            val status = convertToPaymentStatus(statusStr)

            val paymentRequest = io.linkrunner.sdk.models.request.CapturePaymentRequest(
                paymentId = paymentId,
                userId = userId,
                amount = amount,
                type = type,
                status = status
            )

            coroutineScope.launch {
                try {
                    val result = linkrunner.capturePayment(paymentRequest)

                    if (result.isSuccess) {
                        call.resolve()
                    } else {
                        val exception = result.exceptionOrNull()
                        Log.e(TAG, "CapturePayment failed", exception)
                        call.reject("CAPTURE_PAYMENT_FAILED", exception?.message ?: "CapturePayment failed")
                    }
                } catch (e: Exception) {
                    Log.e(TAG, "CapturePayment failed", e)
                    call.reject("CAPTURE_PAYMENT_FAILED", e.message ?: "CapturePayment failed")
                }
            }
        } catch (e: Exception) {
            Log.e(TAG, "Failed to convert payment data", e)
            call.reject("DATA_CONVERSION_FAILED", e.message ?: "Failed to convert payment data")
        }
    }

    /**
     * Remove a payment event
     */
    @PluginMethod
    fun removePayment(call: PluginCall) {
        val userId = call.getString("userId")
        val paymentId = call.getString("paymentId")

        if (userId.isNullOrBlank()) {
            call.reject("INVALID_PARAMETER", "userId is required")
            return
        }

        try {
            val removeRequest = io.linkrunner.sdk.models.request.RemovePaymentRequest(
                paymentId = paymentId,
                userId = userId
            )

            coroutineScope.launch {
                try {
                    val result = linkrunner.removePayment(removeRequest)

                    if (result.isSuccess) {
                        call.resolve()
                    } else {
                        val exception = result.exceptionOrNull()
                        Log.e(TAG, "RemovePayment failed", exception)
                        call.reject("REMOVE_PAYMENT_FAILED", exception?.message ?: "RemovePayment failed")
                    }
                } catch (e: Exception) {
                    Log.e(TAG, "RemovePayment failed", e)
                    call.reject("REMOVE_PAYMENT_FAILED", e.message ?: "RemovePayment failed")
                }
            }
        } catch (e: Exception) {
            Log.e(TAG, "Failed to convert payment removal data", e)
            call.reject("DATA_CONVERSION_FAILED", e.message ?: "Failed to convert payment removal data")
        }
    }

    /**
     * Convert string to PaymentType enum
     */
    private fun convertToPaymentType(typeStr: String?): io.linkrunner.sdk.models.PaymentType {
        return when (typeStr?.uppercase()) {
            "FIRST_PAYMENT" -> io.linkrunner.sdk.models.PaymentType.FIRST_PAYMENT
            "WALLET_TOPUP" -> io.linkrunner.sdk.models.PaymentType.WALLET_TOPUP
            "FUNDS_WITHDRAWAL" -> io.linkrunner.sdk.models.PaymentType.FUNDS_WITHDRAWAL
            "SUBSCRIPTION_CREATED" -> io.linkrunner.sdk.models.PaymentType.SUBSCRIPTION_CREATED
            "SUBSCRIPTION_RENEWED" -> io.linkrunner.sdk.models.PaymentType.SUBSCRIPTION_RENEWED
            "ONE_TIME" -> io.linkrunner.sdk.models.PaymentType.ONE_TIME
            "RECURRING" -> io.linkrunner.sdk.models.PaymentType.RECURRING
            else -> io.linkrunner.sdk.models.PaymentType.DEFAULT
        }
    }

    /**
     * Convert string to PaymentStatus enum
     */
    private fun convertToPaymentStatus(statusStr: String?): io.linkrunner.sdk.models.PaymentStatus {
        return when (statusStr?.uppercase()) {
            "PAYMENT_INITIATED" -> io.linkrunner.sdk.models.PaymentStatus.PAYMENT_INITIATED
            "PAYMENT_FAILED" -> io.linkrunner.sdk.models.PaymentStatus.PAYMENT_FAILED
            "PAYMENT_CANCELLED" -> io.linkrunner.sdk.models.PaymentStatus.PAYMENT_CANCELLED
            else -> io.linkrunner.sdk.models.PaymentStatus.PAYMENT_COMPLETED
        }
    }

    /**
     * Track a custom event
     */
    @PluginMethod
    fun trackEvent(call: PluginCall) {
        val eventName = call.getString("eventName")
        val eventData = call.getObject("eventData")
        val eventId = call.getString("eventId")

        if (eventName.isNullOrBlank()) {
            call.reject("INVALID_PARAMETER", "eventName is required")
            return
        }

        try {
            val eventDataMap = eventData?.let { convertJSObjectToMap(it) }

            coroutineScope.launch {
                try {
                    val result = linkrunner.trackEvent(
                        eventName = eventName,
                        eventData = eventDataMap,
                        eventId = eventId
                    )

                    if (result.isSuccess) {
                        call.resolve()
                    } else {
                        val exception = result.exceptionOrNull()
                        Log.e(TAG, "TrackEvent failed", exception)
                        call.reject("TRACK_EVENT_FAILED", exception?.message ?: "TrackEvent failed")
                    }
                } catch (e: Exception) {
                    Log.e(TAG, "TrackEvent failed", e)
                    call.reject("TRACK_EVENT_FAILED", e.message ?: "TrackEvent failed")
                }
            }
        } catch (e: Exception) {
            Log.e(TAG, "Failed to convert event data", e)
            call.reject("DATA_CONVERSION_FAILED", e.message ?: "Failed to convert event data")
        }
    }

    /**
     * Get attribution data
     */
    @PluginMethod
    fun getAttributionData(call: PluginCall) {
        coroutineScope.launch {
            try {
                val result = linkrunner.getAttributionData()

                if (result.isSuccess) {
                    val attributionData = result.getOrNull()
                    if (attributionData != null) {
                        val response = convertAttributionDataToJSObject(attributionData)
                        call.resolve(response)
                    } else {
                        call.reject("GET_ATTRIBUTION_DATA_FAILED", "Attribution data is null")
                    }
                } else {
                    val exception = result.exceptionOrNull()
                    Log.e(TAG, "GetAttributionData failed", exception)
                    call.reject("GET_ATTRIBUTION_DATA_FAILED", exception?.message ?: "GetAttributionData failed")
                }
            } catch (e: Exception) {
                Log.e(TAG, "GetAttributionData failed", e)
                call.reject("GET_ATTRIBUTION_DATA_FAILED", e.message ?: "GetAttributionData failed")
            }
        }
    }

    /**
     * Set additional integration data
     */
    @PluginMethod
    fun setAdditionalData(call: PluginCall) {
        val integrationDataObj = call.getObject("integrationData")

        if (integrationDataObj == null) {
            call.reject("INVALID_PARAMETER", "integrationData is required")
            return
        }

        try {
            val integrationData = convertToIntegrationData(integrationDataObj)

            if (integrationData.isEmpty()) {
                call.reject("INVALID_PARAMETER", "integrationData cannot be empty")
                return
            }

            coroutineScope.launch {
                try {
                    val result = linkrunner.setAdditionalData(integrationData)

                    if (result.isSuccess) {
                        call.resolve()
                    } else {
                        val exception = result.exceptionOrNull()
                        Log.e(TAG, "SetAdditionalData failed", exception)
                        call.reject("SET_ADDITIONAL_DATA_FAILED", exception?.message ?: "SetAdditionalData failed")
                    }
                } catch (e: Exception) {
                    Log.e(TAG, "SetAdditionalData failed", e)
                    call.reject("SET_ADDITIONAL_DATA_FAILED", e.message ?: "SetAdditionalData failed")
                }
            }
        } catch (e: Exception) {
            Log.e(TAG, "Failed to convert integration data", e)
            call.reject("DATA_CONVERSION_FAILED", e.message ?: "Failed to convert integration data")
        }
    }

    /**
     * Enable or disable PII hashing
     */
    @PluginMethod
    fun enablePIIHashing(call: PluginCall) {
        val enabled = call.getBoolean("enabled", true) ?: true

        try {
            linkrunner.enablePIIHashing(enabled)
            call.resolve()
        } catch (e: Exception) {
            Log.e(TAG, "EnablePIIHashing failed", e)
            call.reject("ENABLE_PII_HASHING_FAILED", e.message ?: "EnablePIIHashing failed")
        }
    }

    /**
     * Get the package version
     */
    @PluginMethod
    fun getPackageVersion(call: PluginCall) {
        try {
            // Read version from package.json in assets or return a default
            val version = "0.0.1" // This will be replaced with actual version reading logic
            val result = JSObject()
            result.put("version", version)
            call.resolve(result)
        } catch (e: Exception) {
            Log.e(TAG, "GetPackageVersion failed", e)
            call.reject("GET_PACKAGE_VERSION_FAILED", e.message ?: "GetPackageVersion failed")
        }
    }

    /**
     * Convert AttributionData to JSObject
     */
    private fun convertAttributionDataToJSObject(attributionData: io.linkrunner.sdk.models.response.AttributionData): JSObject {
        val result = JSObject()
        val data = JSObject()

        // Add deeplink if present
        attributionData.deeplink?.let {
            data.put("deeplink", it)
        }

        // Convert campaign data
        val campaignData = JSObject()
        campaignData.put("id", attributionData.campaignData.id)
        campaignData.put("name", attributionData.campaignData.name)
        campaignData.put("type", attributionData.campaignData.type)
        campaignData.put("installedAt", attributionData.campaignData.installedAt)
        campaignData.put("groupName", attributionData.campaignData.groupName)
        campaignData.put("assetName", attributionData.campaignData.assetName)
        campaignData.put("assetGroupName", attributionData.campaignData.assetGroupName)

        // Add optional fields
        attributionData.campaignData.adNetwork?.let {
            campaignData.put("adNetwork", it)
        }
        attributionData.campaignData.storeClickAt?.let {
            campaignData.put("storeClickAt", it)
        }

        data.put("campaignData", campaignData)
        result.put("data", data)

        return result
    }

    /**
     * Convert JSObject to IntegrationData
     */
    private fun convertToIntegrationData(jsObject: JSObject): io.linkrunner.sdk.models.IntegrationData {
        return io.linkrunner.sdk.models.IntegrationData(
            clevertapId = jsObject.getString("clevertap_id")
        )
    }
}

File: ios/Sources/LinkrunnerPlugin/LinkrunnerPlugin.swift
import Foundation
import Capacitor
import LinkrunnerKit

/**
 * Capacitor plugin for Linkrunner SDK
 * Provides a bridge between JavaScript/TypeScript and the native iOS Linkrunner SDK
 */
@objc(LinkrunnerPlugin)
public class LinkrunnerPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "LinkrunnerPlugin"
    public let jsName = "Linkrunner"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "init", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "signup", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setUserData", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "capturePayment", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "removePayment", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "trackEvent", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getAttributionData", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setAdditionalData", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "enablePIIHashing", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "handleDeeplink", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getPackageVersion", returnType: CAPPluginReturnPromise)
    ]
    
    private var linkrunner: LinkrunnerSDK {
        return LinkrunnerSDK.shared
    }
    
    // MARK: - Initialization
    
    /**
     * Initialize the Linkrunner SDK
     */
    @objc func `init`(_ call: CAPPluginCall) {
        guard let token = call.getString("token"), !token.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            call.reject("Token is required")
            return
        }
        
        let secretKey = call.getString("secretKey")
        let keyId = call.getString("keyId")
        let disableIdfa = call.getBool("disableIdfa") ?? false
        let debug = call.getBool("debug") ?? false
        
        // Initialize the SDK asynchronously
        Task {
            if #available(iOS 15.0, *) {
                await linkrunner.initialize(
                    token: token,
                    secretKey: secretKey,
                    keyId: keyId,
                    disableIdfa: disableIdfa,
                    debug: debug
                )
                call.resolve()
            } else {
                call.reject("iOS 15.0 or later is required")
            }
        }
    }
    
    // MARK: - User Management
    
    /**
     * Register a new user signup
     */
    @objc func signup(_ call: CAPPluginCall) {
        guard let userDataObj = call.getObject("userData") else {
            call.reject("userData is required")
            return
        }
        
        do {
            let userData = try convertToUserData(userDataObj)
            let additionalData = call.getObject("data")
            let additionalDataDict = additionalData != nil ? convertJSObjectToDict(additionalData!) : nil
            
            Task {
                if #available(iOS 15.0, *) {
                    await linkrunner.signup(userData: userData, additionalData: additionalDataDict)
                    call.resolve()
                } else {
                    call.reject("iOS 15.0 or later is required")
                }
            }
        } catch {
            call.reject("Failed to convert user data: \(error.localizedDescription)")
        }
    }
    
    /**
     * Update user data
     */
    @objc func setUserData(_ call: CAPPluginCall) {
        guard let userDataObj = call.getObject("userData") else {
            call.reject("userData is required")
            return
        }
        
        do {
            let userData = try convertToUserData(userDataObj)
            
            Task {
                if #available(iOS 15.0, *) {
                    await linkrunner.setUserData(userData)
                    call.resolve()
                } else {
                    call.reject("iOS 15.0 or later is required")
                }
            }
        } catch {
            call.reject("Failed to convert user data: \(error.localizedDescription)")
        }
    }
    
    // MARK: - Payment Tracking
    
    /**
     * Capture a payment event
     */
    @objc func capturePayment(_ call: CAPPluginCall) {
        guard let userId = call.getString("userId"), !userId.isEmpty else {
            call.reject("userId is required")
            return
        }
        
        guard let amount = call.getDouble("amount") else {
            call.reject("amount is required")
            return
        }
        
        let paymentId = call.getString("paymentId")
        let typeStr = call.getString("type") ?? "DEFAULT"
        let statusStr = call.getString("status") ?? "PAYMENT_COMPLETED"
        
        let type = convertToPaymentType(typeStr)
        let status = convertToPaymentStatus(statusStr)
        
        Task {
            if #available(iOS 15.0, *) {
                await linkrunner.capturePayment(
                    amount: amount,
                    userId: userId,
                    paymentId: paymentId,
                    type: type,
                    status: status
                )
                call.resolve()
            } else {
                call.reject("iOS 15.0 or later is required")
            }
        }
    }
    
    /**
     * Remove a payment event
     */
    @objc func removePayment(_ call: CAPPluginCall) {
        guard let userId = call.getString("userId"), !userId.isEmpty else {
            call.reject("userId is required")
            return
        }
        
        let paymentId = call.getString("paymentId")
        
        Task {
            if #available(iOS 15.0, *) {
                await linkrunner.removePayment(userId: userId, paymentId: paymentId)
                call.resolve()
            } else {
                call.reject("iOS 15.0 or later is required")
            }
        }
    }
    
    // MARK: - Event Tracking and Attribution
    
    /**
     * Track a custom event
     */
    @objc func trackEvent(_ call: CAPPluginCall) {
        guard let eventName = call.getString("eventName"), !eventName.isEmpty else {
            call.reject("eventName is required")
            return
        }
        
        let eventData = call.getObject("eventData")
        let eventDataDict = eventData != nil ? convertJSObjectToDict(eventData!) : nil
        let eventId = call.getString("eventId")
        
        Task {
            if #available(iOS 15.0, *) {
                await linkrunner.trackEvent(
                    eventName: eventName,
                    eventData: eventDataDict,
                    eventId: eventId
                )
                call.resolve()
            } else {
                call.reject("iOS 15.0 or later is required")
            }
        }
    }
    
    /**
     * Get attribution data
     */
    @objc func getAttributionData(_ call: CAPPluginCall) {
        Task {
            if #available(iOS 15.0, *) {
                let attributionData = await linkrunner.getAttributionData()
                let result = convertAttributionDataToJSObject(attributionData)
                call.resolve(result)
            } else {
                call.reject("iOS 15.0 or later is required")
            }
        }
    }
    
    /**
     * Set additional integration data
     */
    @objc func setAdditionalData(_ call: CAPPluginCall) {
        guard let integrationDataObj = call.getObject("integrationData") else {
            call.reject("integrationData is required")
            return
        }
        
        let integrationData = convertToIntegrationData(integrationDataObj)
        
        Task {
            if #available(iOS 15.0, *) {
                await linkrunner.setAdditionalData(integrationData)
                call.resolve()
            } else {
                call.reject("iOS 15.0 or later is required")
            }
        }
    }
    
    /**
     * Enable or disable PII hashing
     */
    @objc func enablePIIHashing(_ call: CAPPluginCall) {
        let enabled = call.getBool("enabled") ?? true
        
        if #available(iOS 15.0, *) {
            linkrunner.enablePIIHashing(enabled)
            call.resolve()
        } else {
            call.reject("iOS 15.0 or later is required")
        }
    }
    
    /**
     * Handle a deeplink for re-engagement attribution
     * Call this method when the app is opened via a deeplink, regardless of app state
     */
    @objc func handleDeeplink(_ call: CAPPluginCall) {
        let deeplinkUrl = call.getString("deeplinkUrl")
        
        // Handle empty URLs gracefully (matches SDK behavior)
        if deeplinkUrl == nil || deeplinkUrl!.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            call.resolve()
            return
        }
        
        Task {
            if #available(iOS 15.0, *) {
                await linkrunner.handleDeeplink(url: deeplinkUrl)
                call.resolve()
            } else {
                call.reject("iOS 15.0 or later is required")
            }
        }
    }
    
    /**
     * Get the package version
     */
    @objc func getPackageVersion(_ call: CAPPluginCall) {
        // Read version from package.json or return a default
        let version = "0.0.1" // This will be replaced with actual version reading logic
        call.resolve(["version": version])
    }
}


// MARK: - Type Conversion Helpers

extension LinkrunnerPlugin {
    /**
     * Convert JSObject to UserData
     */
    private func convertToUserData(_ jsObject: JSObject) throws -> UserData {
        guard let id = jsObject["id"] as? String, !id.isEmpty else {
            throw NSError(domain: "LinkrunnerPlugin", code: 1, userInfo: [NSLocalizedDescriptionKey: "User id is required"])
        }
        
        return UserData(
            id: id,
            name: jsObject["name"] as? String,
            phone: jsObject["phone"] as? String,
            email: jsObject["email"] as? String,
            isFirstTimeUser: jsObject["is_first_time_user"] as? Bool,
            userCreatedAt: jsObject["user_created_at"] as? String,
            mixPanelDistinctId: jsObject["mixpanel_distinct_id"] as? String,
            amplitudeDeviceId: jsObject["amplitude_device_id"] as? String,
            posthogDistinctId: jsObject["posthog_distinct_id"] as? String,
            brazeDeviceId: jsObject["braze_device_id"] as? String,
            gaAppInstanceId: jsObject["ga_app_instance_id"] as? String,
            gaSessionId: jsObject["ga_session_id"] as? String,
            netcoreDeviceGuid: jsObject["netcore_device_guid"] as? String
        )
    }
    
    /**
     * Convert JSObject to dictionary
     */
    private func convertJSObjectToDict(_ jsObject: JSObject) -> SendableDictionary {
        var dict: SendableDictionary = [:]
        for (key, value) in jsObject {
            dict[key] = value
        }
        return dict
    }
    
    /**
     * Convert string to PaymentType enum
     */
    private func convertToPaymentType(_ typeStr: String) -> PaymentType {
        switch typeStr.uppercased() {
        case "FIRST_PAYMENT":
            return .firstPayment
        case "WALLET_TOPUP":
            return .walletTopup
        case "FUNDS_WITHDRAWAL":
            return .fundsWithdrawal
        case "SUBSCRIPTION_CREATED":
            return .subscriptionCreated
        case "SUBSCRIPTION_RENEWED":
            return .subscriptionRenewed
        case "ONE_TIME":
            return .oneTime
        case "RECURRING":
            return .recurring
        default:
            return .default
        }
    }
    
    /**
     * Convert string to PaymentStatus enum
     */
    private func convertToPaymentStatus(_ statusStr: String) -> PaymentStatus {
        switch statusStr.uppercased() {
        case "PAYMENT_INITIATED":
            return .initiated
        case "PAYMENT_FAILED":
            return .failed
        case "PAYMENT_CANCELLED":
            return .cancelled
        default:
            return .completed
        }
    }
    
    /**
     * Convert JSObject to IntegrationData
     */
    private func convertToIntegrationData(_ jsObject: JSObject) -> IntegrationData {
        return IntegrationData(
            clevertapId: jsObject["clevertapId"] as? String
        )
    }
    
    /**
     * Convert AttributionData to JSObject
     */
    private func convertAttributionDataToJSObject(_ attributionData: LRAttributionDataResponse) -> JSObject {
        var result: JSObject = [:]
        var data: JSObject = [:]
        
        // Add deeplink if present
        if let deeplink = attributionData.deeplink {
            data["deeplink"] = deeplink
        }
        
        // Convert campaign data if present
        if let campaignData = attributionData.campaignData {
            var campaignDict: JSObject = [:]
            campaignDict["id"] = campaignData.id
            campaignDict["name"] = campaignData.name
            campaignDict["type"] = campaignData.type.rawValue
            
            // Add optional fields
            if let adNetwork = campaignData.adNetwork {
                campaignDict["adNetwork"] = adNetwork.rawValue
            }
            if let groupName = campaignData.groupName {
                campaignDict["groupName"] = groupName
            }
            if let assetName = campaignData.assetName {
                campaignDict["assetName"] = assetName
            }
            if let assetGroupName = campaignData.assetGroupName {
                campaignDict["assetGroupName"] = assetGroupName
            }
            
            // Format dates
            let dateFormatter = ISO8601DateFormatter()
            if let installedAt = campaignData.installedAt {
                campaignDict["installedAt"] = dateFormatter.string(from: installedAt)
            }
            if let storeClickAt = campaignData.storeClickAt {
                campaignDict["storeClickAt"] = dateFormatter.string(from: storeClickAt)
            }
            
            data["campaignData"] = campaignDict
        }
        
        result["data"] = data
        return result
    }
}
